test/expect-tests/fiber/fiber_tests.ml

```diff
@@ -847,6 +847,60 @@ let%expect_test "raise exception" =
        (fun () -> Pool.stop pool));
   [%expect {| Caught Exit |}]
 
+let%expect_test "double run a pool" =
+  (* Calling [Pool.run] on the same pool shouldn't be allowed *)
+  (Scheduler.run
+  @@
+  let pool = Pool.create () in
+  Fiber.fork_and_join_unit (fun () -> Pool.run pool) (fun () -> Pool.run pool));
+  [%expect.unreachable]
+  [@@expect.uncaught_exn {| (Test_scheduler.Never) |}]
+
+let%expect_test "start -> stop -> start a pool" =
+  (* We shouldn't be able to call [Pool.run] after stopping it *)
+  (Scheduler.run
+  @@
+  let pool = Pool.create () in
+  let* () =
+    Fiber.fork_and_join_unit
+      (fun () -> Pool.run pool)
+      (fun () -> Pool.stop pool)
+  in
+  Pool.run pool);
+  [%expect.unreachable]
+  [@@expect.uncaught_exn {| (Test_scheduler.Never) |}]
+
+let%expect_test "stop a pool and then run it" =
+  (* Weird deadlock when we call [stop] after [run] *)
+  (Scheduler.run
+  @@
+  let pool = Pool.create () in
+  let* () = Pool.stop pool in
+  let* () = Pool.run pool in
+  Pool.run pool);
+  [%expect.unreachable]
+  [@@expect.uncaught_exn {| (Test_scheduler.Never) |}]
+
+let%expect_test "pool - weird deadlock" =
```

Comment:
@snowleopard this one wasted a lot of my time when implementing the action runner. Let's see if you know why this happens offhand :)

Reply:
I have no clue what's happening here, so no guesses from me!

Reply:
The issue is the underlying mvar implementation. Writing to an mvar that is empty does not block, so the first test happily succeeds. The second test fails because the second `Pool.task` blocks on writing to the mvar.

All of this makes sense when dealing with mvars - but pools aren't mvars. Pools are just used to "reap" sets of fibers, that is wait for them to complete and handle any of their exceptions if need be. So it doesn't make sense for them to inherit this behavior at all.